### PR TITLE
Displace-tracking development in heavy ion tracking 

### PR DESCRIPTION
**Pull request by KongTu**
This displace-tracking development in heavy ion reconstruction is to find (like CMS standard pp tracking) the large impact parameter tracks for strange particles or heavy-flavor particles reconstruction. 

This PR contains the following items:
- Added the configurations for 3 iterations: "hiMixedTripletStep", "hiPixelLessStep", "hiTobTecStep". Modified on top of the default tracking algorithm from standard pp tracking. 
- The TrackingRegion parameters are modified to reduce the timing on these 3 iterations above, e.g., the minPt and originHalfLength. 
- A "scaling" function has been implemented (in this file GlobalTrackingRegionWithVerticesProducer.h), where according to the number of pixel clusters (mostly for more central heavy ion events), the TrackingRegion parameters, e.g., minPt, originRadius, originHalfLength, are scaled up (pt) or down (originRaidu or originHalfLength) linearly above a certain threshold (for example Npix=20000). Therefore the timing can be greatly reduced towards very central events. 

Supporting presentations:
- Performance plots on efficiency, fake rate, secondary rate (where it matters), https://indico.cern.ch/event/699094/contributions/2867255/attachments/1587914/2512006/Tracking_PR_preparation.pdf
Note that the comparison here is to compare "default" values from the standard pp tracking vs the modified values in the TrackingRegion parameters of the added iterations. 

- Study of V0s in terms of added 3 iterations:
https://indico.cern.ch/event/638128/contributions/2585615/attachments/1460150/2255756/Tracking_6_90X.pdf
https://indico.cern.ch/event/646309/contributions/2625076/attachments/1475706/2285788/Tracking_7_90X.pdf

- Timing study as a function of Npix, https://indico.cern.ch/event/699094/contributions/2867255/attachments/1587914/2511992/Timing_displacedTracking_Jan24.pdf

@abaty @mandrenguyen

### CODE REVIEW

**Comment by slava77** [comment]
@KongTu 

> Timing study as a function of Npix, https://indico.cern.ch/event/699094/contributions/2867255/attachments/1587914/2511992/Timing_displacedTracking_Jan24.pdf

the plot has "blue" (this PR) both above and below the baseline.
Please clarify what is happening in the lower bins.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by abaty** [comment]
@slava77 @KongTu 
> the plot has "blue" (this PR) both above and below the baseline.
Please clarify what is happening in the lower bins.

Each point here represents the processing time of a single event, so I suspect the difference seen in the lower bins is just a fluctuation in the processing time.  The code here does not 'remove' any features for central events to make them run faster; it only adds tracking iterations for more peripheral events.

**Comment by slava77** [comment]
@cmsbuild please test workflow 148,150

**Comment by slava77** [comment]
I have some general comments to the PR:
- I was informed earlier that the strategy for upcoming HI reco developments is to effectively merge with pp and apply era modifications on top of it for AA. This PR seems to make the tracking look more similar to the pp setup but from a different direction. Please clarify the scope and the plans and why we can't just advance with the pp on AA.
- (assuming the option above is a no-go) I see many additions in configs that look like copy-paste from the RecoTracker/IterativeTracking with sometimes only "hi" prefix added. It seems more practical to simply import the modules and sequences from RecoTracker/IterativeTracking without changing the names and just modify the few parameters that have to be different and use them.

**Comment by KongTu** [comment]
- The idea of this development is to get a working version of displace tracking in HI tracking sequence and it is not 100% sure at the moment that the approach of modifying the pp tracking would indeed work for AA events. This can be regarded as a backup tracking solution in case the our main strategy turns out not suitable. 

- The naming of those added iterations were to follow the naming convention of those early iteration in HI tracking. It seems to me that mixing the naming in HI tracking is not the best way of doing it. Otherwise, some of the early iterations probably should also change the name.

**Comment by abaty** [comment]
@slava77 @KongTu 

> I was informed earlier that the strategy for upcoming HI reco developments is to effectively merge with pp and apply era modifications on top of it for AA. 

Just to clarifying a bit more on the strategy, we are currently investigating the option you mention above.  However, whether the decision to use the pp or HI tracking has not been taken yet and will probably strongly depend on the computing resources available, expected data volumes, and the final timing of both strategies.  These are all not finalized yet as far as I am aware.

**Comment by slava77** [comment]
On 1/25/18 12:36 PM, KongTu wrote:
>   *
> 
>     The idea of this development is to get a working version of displace
>     tracking in HI tracking sequence and it is not 100% sure at the
>     moment that the approach of modifying the pp tracking would indeed
>     work for AA events. This can be regarded as a backup tracking
>     solution in case the our main strategy turns out not suitable.

OK.

> 
>   *
> 
>     The naming of those added iterations were to follow the naming
>     convention of those early iteration in HI tracking. It seems to me
>     that mixing the naming in HI tracking is not the best way of doing
>     it. Otherwise, some of the early iterations probably should also
>     change the name.

With the recent expansion, I think that the approach to copy and prepend
"hi"
is not really sustainable.
Are there any functional arguments to do this other than the historical
naming convention used for modules in the past?

The goal of converging to a modified pp setup seems much more achievable
by reusing the elements from pp from the start.

In any case, if new modules are unavoidable, they better be cloned
directly from the pp iter tracking equivalents replacing only the
parameters that should be different.
This is again with the target of modified pp in mind.

This should reduce this PR by probably a factor of 2, while the option
reusing the names of pp modules will reduce it by probably another
factor of 2 and make it easier to maintain.

**Comment by slava77** [comment]
@KongTu @abaty 
I just wanted to check on the status or plans for this PR.

**Comment by KongTu** [comment]
@slava77 
We are working on to get a shorter version as suggested.

**Comment by slava77** [comment]
@cmsbuild please test workflow 148,150

**Comment by KongTu** [comment]
@slava77 @abaty 
We have updated to a shorter version as suggested. Please check. Thanks!

File: RecoHI/HiTracking/python/MergeTrackCollectionsHI_cff.py [hunk 1]
```diff
@@ -86,6 +86,9 @@
                       cms.InputTag('hiDetachedTripletStepTracks'),
                       cms.InputTag('hiLowPtTripletStepTracks'),
                       cms.InputTag('hiPixelPairGlobalPrimTracks'),
+                      cms.InputTag('hiMixedTripletStepTracks'),
+		      cms.InputTag('hiPixelLessStepTracks'),
+		      cms.InputTag('hiTobTecStepTracks'),
```
Comment (slava77) [inline RecoHI/HiTracking/python/MergeTrackCollectionsHI_cff.py]: the indentation is inconsistent

Comment (KongTu) [reply RecoHI/HiTracking/python/MergeTrackCollectionsHI_cff.py]: Seems like a bug on git to diff the code. The indentation is consistent if you click into the file, https://github.com/cmsHiTracking/cmssw/blob/bb4fa82ee7144187e8b8da40039ab680ac64607c/RecoHI/HiTracking/python/MergeTrackCollectionsHI_cff.py

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py [hunk 2]
```diff
@@ -1,206 +1,159 @@
...
-
-                                        
-
+from RecoTracker.IterativeTracking.MixedTripletStep_cff import *
+from HIPixelTripletSeeds_cff import *
+from HIPixel3PrimTracks_cfi import *
+
+#cluster remover
+hiMixedTripletStepClusters = cms.EDProducer("HITrackClusterRemover",
+     clusterLessSolution = cms.bool(True),
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: oldClusterRemovalInfo was removed.
It's present in the non-HI version.
Was this removal intentional?

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py [hunk 3]
```diff
@@ -1,206 +1,159 @@
...
-hiMixedTripletSeedsA.RegionFactoryPSet.RegionPSet.ptMin = 4.0
-hiMixedTripletSeedsA.RegionFactoryPSet.RegionPSet.originRadius = 0.005
-#hiMixedTripletSeedsA.RegionFactoryPSet.RegionPSet.nSigmaZ = 4.0
-hiMixedTripletSeedsA.RegionFactoryPSet.RegionPSet.originHalfLength = 10.0
-
-hiMixedTripletSeedsA.OrderedHitsFactoryPSet.GeneratorPSet.maxElement = 5000000
-hiMixedTripletSeedsA.ClusterCheckPSet.MaxNumberOfPixelClusters = 5000000
-hiMixedTripletSeedsA.ClusterCheckPSet.MaxNumberOfCosmicClusters = 50000000
+mixedTripletStepSeedLayersA.layerList = cms.vstring('BPix1+BPix2+BPix3', 'BPix1+BPix2+FPix1_pos', 'BPix1+BPix2+FPix1_neg', 
+                                        'BPix1+FPix1_pos+FPix2_pos', 'BPix1+FPix1_neg+FPix2_neg',
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: are these really not picked up in the previous triplet iterations?

File: RecoTracker/TkTrackingRegions/python/GlobalTrackingRegionWithVertices_cfi.py [hunk 4]
```diff
@@ -14,7 +14,19 @@
         useFoundVertices = cms.bool(True),
         useFakeVertices = cms.bool(False),
         maxNVertices = cms.int32(-1),
-        nSigmaZ = cms.double(4.0)
+        nSigmaZ = cms.double(4.0),
+
+        #parameters for HI collisions; doesn't do anything if all 3 booleans are False
+       
+        originRScaling4BigEvts = cms.bool(False),
```
Comment (slava77) [inline RecoTracker/TkTrackingRegions/python/GlobalTrackingRegionWithVertices_cfi.py]: @makortel 
is this file still supposed to be used?
It's pretty confusing to have the auto-generated RecoTracker/TkTrackingRegions/globalTrackingRegionWithVertices_cfi.py and this file.

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py [hunk 5]
```diff
@@ -1,206 +1,159 @@
...
+     ptMin = 0.4,
+     originRadius = 1.5,
+     originRScaling4BigEvts = cms.bool(True),
+     halfLengthScaling4BigEvts = cms.bool(False),
+     ptMinScaling4BigEvts = cms.bool(True),
+     minOriginR = 0,
+     minHalfLength = 0,
+     maxPtMin = 0.7,
+     scalingStartNPix = 20000,
+     scalingEndNPix = 35000
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: more than a half of the above parameters are the same as in the original globalTrackingRegionWithVertices_cfi and should not really be repeated.
It may also be more practical to define also hiGlobalTrackingRegionWithVertices_cfi so that it can be reused directly instead of repeated copy-pastes of almost 20 lines

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py (hunk 3 above)
Comment (abaty) [reply RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: Although the layer list is the same as previous iterations, the size of the tracking region is much larger.  On line 43 we are using a fixed halfLength of nearly 4 cm, while the previous triplet steps use 0.5 cm (high pt step) and less than 4 sigma from the constraining vertex (low pt step).

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py [hunk 6]
```diff
@@ -1,206 +1,159 @@
...
-import TrackingTools.KalmanUpdators.Chi2MeasurementEstimator_cfi
-hiMixedTripletChi2Est = TrackingTools.KalmanUpdators.Chi2MeasurementEstimator_cfi.Chi2MeasurementEstimator.clone(
-    ComponentName = cms.string('hiMixedTripletChi2Est'),
-    nSigma = cms.double(3.0),
-    MaxChi2 = cms.double(16.0)
-    )
-
+from RecoTracker.IterativeTracking.MixedTripletStep_cff import _mixedTripletStepTrajectoryFilterBase
+_mixedTripletStepTrajectoryFilterBase.minimumNumberOfHits = 5
+_mixedTripletStepTrajectoryFilterBase.minPt = 0.4
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: This does not work (changes are useless the way they are applied), please apply modifications to mixedTripletStepTrajectoryFilter instead of this internal ``` _mixedTripletStepTrajectoryFilterBase```

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py [hunk 7]
```diff
@@ -1,206 +1,159 @@
...
-hiMixedTripletTrajectoryBuilder = RecoTracker.CkfPattern.GroupedCkfTrajectoryBuilder_cfi.GroupedCkfTrajectoryBuilder.clone(
-    MeasurementTrackerName = '',
-    trajectoryFilter = cms.PSet(refToPSet_ = cms.string('hiMixedTripletTrajectoryFilter')),
-    propagatorAlong = cms.string('hiMixedTripletPropagator'),
-    propagatorOpposite = cms.string('hiMixedTripletPropagatorOpposite'),
-    clustersToSkip = cms.InputTag('hiMixedTripletClusters'),
-    maxCand = 2,
-    estimator = cms.string('hiMixedTripletChi2Est')
-    )
+hiMixedTripletStepTracks = mixedTripletStepTracks.clone()
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: please define tracks after the candidates just to have them in a natural sequence.

File: RecoHI/HiTracking/python/hiPixelLessStep_cff.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+from RecoTracker.IterativeTracking.PixelLessStep_cff import *
+from HIPixelTripletSeeds_cff import *
+from HIPixel3PrimTracks_cfi import *
+
+##########################################################################
+# Large impact parameter tracking using TIB/TID/TEC stereo layer seeding #
+##########################################################################
+
+hiPixelLessStepClusters = cms.EDProducer("HITrackClusterRemover",
+     clusterLessSolution = cms.bool(True),
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiPixelLessStep_cff.py]: about 80% of this PSet is identical to the hiMixedTripletStepClusters.
Please define a default and reuse it.

File: RecoHI/HiTracking/python/hiPixelLessStep_cff.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+pixelLessStepSeedLayers.TID.skipClusters   = cms.InputTag('hiPixelLessStepClusters')
+pixelLessStepSeedLayers.MTID.skipClusters   = cms.InputTag('hiPixelLessStepClusters')
+pixelLessStepSeedLayers.TEC.skipClusters   = cms.InputTag('hiPixelLessStepClusters')
+pixelLessStepSeedLayers.MTEC.skipClusters   = cms.InputTag('hiPixelLessStepClusters')
+
+# TrackingRegion
+from RecoTracker.TkTrackingRegions.globalTrackingRegionWithVertices_cfi import globalTrackingRegionWithVertices as _globalTrackingRegionWithVertices
+hiPixelLessStepTrackingRegions = _globalTrackingRegionWithVertices.clone(RegionPSet=dict(
+     precise = True,
+     useMultipleScattering = False,
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiPixelLessStep_cff.py]: same comments here as for hiMixedTripletStepTrackingRegionsA

File: RecoHI/HiTracking/python/hiPixelLessStep_cff.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+))
+
+# seeding
+pixelLessStepHitDoublets.clusterCheck = ""
+pixelLessStepHitDoublets.trackingRegions = "hiPixelLessStepTrackingRegions"
+
+# QUALITY CUTS DURING TRACK BUILDING
+from RecoTracker.IterativeTracking.PixelLessStep_cff import _pixelLessStepTrajectoryFilterBase
+_pixelLessStepTrajectoryFilterBase.minimumNumberOfHits = 5
+_pixelLessStepTrajectoryFilterBase.minPt = 0.7
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiPixelLessStep_cff.py]: this doesn't work either.
Please modify the "final" object ``` pixelLessStepTrajectoryFilter ```

File: RecoHI/HiTracking/python/hiPixelLessStep_cff.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+                             pixelLessStepSeedLayers*
+                             hiPixelLessStepTrackingRegions*
+                             pixelLessStepHitDoublets*
+                             pixelLessStepHitTriplets*
+                             pixelLessStepSeeds*
+                             pixelLessStepTrackCandidates*
+                             hiPixelLessStepTracks*
+                             hiPixelLessStepSelector*
+                             hiPixelLessStepQual
+                             )
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiPixelLessStep_cff.py]: a newline would be nice

File: RecoHI/HiTracking/python/hiTobTecStep_cff.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+# Pair seeds
+tobTecStepHitDoubletsPair.clusterCheck = ""
+tobTecStepHitDoubletsPair.trackingRegions = "hiTobTecStepTrackingRegionsPair"
+
+
+# QUALITY CUTS DURING TRACK BUILDING (for inwardss and outwards track building steps)
+from RecoTracker.IterativeTracking.TobTecStep_cff import _tobTecStepTrajectoryFilterBase
+_tobTecStepTrajectoryFilterBase.minimumNumberOfHits = 5
+_tobTecStepTrajectoryFilterBase.minPt = 0.85
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiTobTecStep_cff.py]: essentially the same comments apply to this file as for hiPixelLessStep_cff.py

File: RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h [hunk 13]
```diff
@@ -91,13 +112,56 @@ class GlobalTrackingRegionWithVerticesProducer : public TrackingRegionProducer
...
+              if(thePtMinScaling)      scaledPtMin = theMaxPtMin; 
+            }
+            //second condition - scale radius linearly by Npix in the region from ScalingStart to ScalingEnd
+            else if((nPix <= theScalingEnd) && (nPix > theScalingStart)){
+              if(theOriginRScaling) scaledOriginRadius = theOriginRadius - (theOriginRadius-theMinOriginR)*(nPix-theScalingStart)/(theScalingEnd-theScalingStart);
+              if(theHalfLengthScaling) scaledHalfLength = theFixedError - (theFixedError-theMinHalfLength)*(nPix-theScalingStart)/(theScalingEnd-theScalingStart);
+              if(thePtMinScaling) scaledPtMin = thePtMin - (thePtMin-theMaxPtMin)*(nPix-theScalingStart)/(theScalingEnd-theScalingStart);
+            }
+            //std::cout << "NumberOfPixels: " <<  nPix << std::endl;
+            //std::cout << "Scaled Origin R: " << scaledOriginRadius << " Default Origin R: " << theOriginRadius <<" Scaled HalfLength: " << scaledHalfLength << " Default H.L.: " << theFixedError <<" Scaled pT Min: " << scaledPtMin << " Default pT Min: " << thePtMin << std::endl;
```
Comment (slava77) [inline RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h]: please cleanup the commented out debugs.
If they are still needed, better enable them with a config parameter or use LogDebug

File: RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h [hunk 14]
```diff
@@ -128,6 +192,17 @@ class GlobalTrackingRegionWithVerticesProducer : public TrackingRegionProducer
   edm::EDGetTokenT<reco::VertexCollection> 	 token_vertex; 
   edm::EDGetTokenT<reco::BeamSpot> 	 token_beamSpot; 
 
+  //HI-related variables
+  edm::InputTag pixelClustersForScaling;
```
Comment (slava77) [inline RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h]: this parameter is not necessary, all necessary reading functionality is contained in token_pc already

File: RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h [hunk 15]
```diff
@@ -91,13 +112,56 @@ class GlobalTrackingRegionWithVerticesProducer : public TrackingRegionProducer
...
+          if(theOriginRScaling || thePtMinScaling || theHalfLengthScaling){
+            //Use the unscaled radius unless one of the two conditions below is met
+            double scaledOriginRadius = theOriginRadius;
+            double scaledHalfLength   = theFixedError;
+            double scaledPtMin        = thePtMin;
+
+            //calculate nPixels (adapted from TkSeedGenerator/src/ClusterChecker.cc)
+            double nPix = 0;
+            edm::Handle<edmNew::DetSetVector<SiPixelCluster> > pixelClusterDSV;
+            ev.getByToken(token_pc, pixelClusterDSV);
```
Comment (slava77) [inline RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h]: please move this retrieval outside of the loop over vertices.

File: RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h [hunk 16]
```diff
@@ -91,13 +112,56 @@ class GlobalTrackingRegionWithVerticesProducer : public TrackingRegionProducer
...
+            if (!pixelClusterDSV.failedToGet()) {
+              const edmNew::DetSetVector<SiPixelCluster> & input = *pixelClusterDSV;
+              nPix = input.dataSize();
+            }
+            else{
+              edm::LogError("GlobalTrackingRegionProducerFromVertex")<<"could not get any SiPixel cluster collections of type edm::DetSetVector<SiPixelCluster>";
+              nPix = theScalingEnd+1;//if can't find collection, default to minimum radius to be safe
```
Comment (slava77) [inline RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h]: this seems like a symptom of a configuration error.
This should be resolved with an exception, not a LogError, which is specific for a given event.

Comment (slava77) [reply RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h]: the easiest here is just to remove the if-else and leave just the nPix computation.

File: RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h [hunk 17]
```diff
@@ -91,13 +112,56 @@ class GlobalTrackingRegionWithVerticesProducer : public TrackingRegionProducer
...
+ 
+            //first condition is for high occupancy, second makes sure we won't divide by zero or a negative number
+            if((nPix > theScalingEnd) || ((theScalingEnd-theScalingStart) <= 0)){
+              if(theOriginRScaling)    scaledOriginRadius = theMinOriginR;   // sets parameters to minimum value from PSet
+              if(theHalfLengthScaling) scaledHalfLength = theMinHalfLength;
+              if(thePtMinScaling)      scaledPtMin = theMaxPtMin; 
+            }
+            //second condition - scale radius linearly by Npix in the region from ScalingStart to ScalingEnd
+            else if((nPix <= theScalingEnd) && (nPix > theScalingStart)){
+              if(theOriginRScaling) scaledOriginRadius = theOriginRadius - (theOriginRadius-theMinOriginR)*(nPix-theScalingStart)/(theScalingEnd-theScalingStart);
```
Comment (slava77) [inline RecoTracker/TkTrackingRegions/plugins/GlobalTrackingRegionWithVerticesProducer.h]: please define a temporary instead of repeated re-computation, smth like ```auto aScale = (nPix-theScalingStart)/(theScalingEnd-theScalingStart)```

File: RecoTracker/TkTrackingRegions/python/GlobalTrackingRegionWithVertices_cfi.py (hunk 4 above)
Comment (slava77) [reply RecoTracker/TkTrackingRegions/python/GlobalTrackingRegionWithVertices_cfi.py]: @VinInn please note this "confusing" file as well.

**Comment by slava77** [comment]
@KongTu 
I'm done with the visual inspection.
It looks like at least a few of the comments will change the behavior.

File: RecoTracker/TkTrackingRegions/python/GlobalTrackingRegionWithVertices_cfi.py (hunk 4 above)
Comment (makortel) [reply RecoTracker/TkTrackingRegions/python/GlobalTrackingRegionWithVertices_cfi.py]: @slava77 This file is not used in pp tracking (and I was almost writing that it is obsolete), but I see it gets used for <= 2016 HI tracking (at least in sequences including `hiPixelPairStep`) in
https://github.com/cms-sw/cmssw/blob/3b4acba5c56c01824fcce1247605eebf8d61b486/RecoHI/HiTracking/python/hiPixelPairStep_cff.py#L50-L51
via
https://github.com/cms-sw/cmssw/blob/3b4acba5c56c01824fcce1247605eebf8d61b486/RecoTracker/TkSeedGenerator/python/GlobalSeedsFromPairsWithVertices_cff.py#L15-L24

Basically what should be done first is to migrate the phase0 `hiPixelPairStep` to the "new seeding framework".

File: RecoHI/HiTracking/python/MergeTrackCollectionsHI_cff.py (hunk 1 above)
Comment (makortel) [reply RecoHI/HiTracking/python/MergeTrackCollectionsHI_cff.py]: Line 89 is intended with spaces, while lines 90 and 91 have a mixture of spaces and tabs.

**Comment by slava77** [comment]
@KongTu 
please clarify on the status of this PR from your side.
Thank you.

**Comment by KongTu** [comment]
@slava77 Still working on it. Sorry, will do it very soon.

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py (hunk 2 above)
Comment (KongTu) [reply RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: Yes

**Comment by KongTu** [comment]
@slava77 @abaty Changes are made. Please check. Thanks!

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
the failures are at the configuration stage

```
   [1] Constructing module: class=SeedGeneratorFromRegionHitsEDProducer label='hiPixelPairSeeds'
Exception Message:
MissingParameter: Parameter 'originRScaling4BigEvts' not found.

```

**Comment by abaty** [comment]
I have modified GlobalSeedsFromPairsWithVertices_cff.py to pick up the auto-generated cfi file instead of using the confusing config file mentioned in the review comments above.  My understanding is that moving off this will render the confusing file obsolete.

Workflow 140.53 (cause of previous test failure) now also runs to completion.

**Comment by slava77** [comment]
> My understanding is that moving off this will render the confusing file obsolete.

should we dare to make one more step and remove this file (GlobalTrackingRegionWithVertices_cfi), or is it still used somewhere?

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTracker/TkSeedGenerator/python/GlobalSeedsFromPairsWithVertices_cff.py [hunk 18]
```diff
@@ -21,7 +21,7 @@
       maxElement = cms.uint32(1000000)
     ),
     RegionFactoryPSet = cms.PSet(
-      RegionPSetWithVerticesBlock,
+      RegionPSet = globalTrackingRegionWithVertices.RegionPSet,
```
Comment (makortel) [inline RecoTracker/TkSeedGenerator/python/GlobalSeedsFromPairsWithVertices_cff.py]: Please clone the `globalTrackingRegionWithVertices.RegionPSet`. If anyone modifies `globalSeedsFromPairsWithVertices.RegionPSet` the changes get propagated to `globalTrackingRegionWithVertices.RegionPSet` as well.

**Comment by makortel** [comment]
> > My understanding is that moving off this will render the confusing file obsolete.
> 
> should we dare to make one more step and remove this file (GlobalTrackingRegionWithVertices_cfi), or is it still used somewhere?

According to `git grep` `GlobalSeedsFromPairsWithVertices_cff` is the only place importing it, so removal should be safe.

File: RecoTracker/TkSeedGenerator/python/GlobalSeedsFromPairsWithVertices_cff.py [hunk 19]
```diff
@@ -12,7 +12,7 @@
 from RecoTracker.TkSeedingLayers.PixelLayerPairs_cfi import *
 from RecoTracker.TkSeedingLayers.MixedLayerPairs_cfi import *
 
-from RecoTracker.TkTrackingRegions.GlobalTrackingRegionWithVertices_cfi import *
+from RecoTracker.TkTrackingRegions.globalTrackingRegionWithVertices_cfi import *
```
Comment (makortel) [inline RecoTracker/TkSeedGenerator/python/GlobalSeedsFromPairsWithVertices_cff.py]: In principle the proper file to import from is `globalTrackingRegionWithVertices_cff` which has bunch of era customizations on top of the auto-generated one
https://github.com/cms-sw/cmssw/blob/29ff9cba70150b1b66144603a47e376fdfac2183/RecoTracker/TkTrackingRegions/python/globalTrackingRegionWithVertices_cff.py#L3-L20

Maybe we should take a look on how to auto-generate `globalTrackingRegionWithVerticesDefault_cfi` (or similar) to have a file with customizations called `globalTrackingRegionWithVertices_cfi`.

**Comment by abaty** [comment]
I've removed the obsolete file, changed the import to the cff, and make the RegionPSet a clone.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py (hunk 5 above)
Comment (slava77) [reply RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: It looks like a large fraction of the previous comment still applies
- halfLengthScaling4BigEvts has a common value with what's in the default globalTrackingRegionWithVertices_cfi. Was it really meant to be False?
- same for minOriginR
- same for minHalfLength

In case you find it complicated to define hiGlobalTrackingRegionWithVertices_cfi via PSet validation (fillDescriptions), please use the same code pattern as for hiMixedTripletStepClusters: set the "default" values here and then clone from it in other cases with modifications of only parameters that are changing

Comment (abaty) [reply RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: The default parameters are now set in hiMixedTripletStep and then cloned into the other 2 iterations.

**Comment by slava77** [comment]
@cmsbuild please test

@abaty 
thank you for the quick updates

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py [hunk 20]
```diff
@@ -1,206 +1,136 @@
...
-#hiMixedTripletSeedsA.RegionFactoryPSet.RegionPSet.nSigmaZ = 4.0
-hiMixedTripletSeedsA.RegionFactoryPSet.RegionPSet.originHalfLength = 10.0
-
-hiMixedTripletSeedsA.OrderedHitsFactoryPSet.GeneratorPSet.maxElement = 5000000
-hiMixedTripletSeedsA.ClusterCheckPSet.MaxNumberOfPixelClusters = 5000000
-hiMixedTripletSeedsA.ClusterCheckPSet.MaxNumberOfCosmicClusters = 50000000
-
+mixedTripletStepSeedLayersA.layerList = cms.vstring('BPix1+BPix2+BPix3', 'BPix1+BPix2+FPix1_pos', 'BPix1+BPix2+FPix1_neg', 
+                                        'BPix1+FPix1_pos+FPix2_pos', 'BPix1+FPix1_neg+FPix2_neg', 
+                                        'BPix2+FPix1_pos+FPix2_pos', 'BPix2+FPix1_neg+FPix2_neg')
```
Comment (slava77) [inline RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: the modifications made here actually propagate to hiRegitMuMixedTripletStepSeedLayersA.
I'm not sure that this is desired.

Based on the phase-0 configuration (wf 140.53) this is the only place with such dependency.

An easy fix is to set the layerList in hiRegitMuMixedTripletStepSeedLayersA explicitly.
OTOH, perhaps it's OK to have a more covering layer list for hiRegitMuMixedTripletStepSeedLayersA

@KongTu @abaty please check

**Comment by slava77** [comment]
Based on wf 150.0 I see that 
- CPU time in reco increases by about 25% (from 11.4 to 14.2 s/ev)
    - most of the increase is in the new modules +2.7 s/ev
    - the rest is about 0.2 s/ev and it shows up in related modules (e.g. hiGeneralTracks or hiMuons1stStep, and particleFlowBlock); 
- RSS peak (based on MemoryCheck) goes up by about 100 MB. This looks ~OK and should likely be smaller on a job writing AOD

Physics performance roughly looks as expected:
- there are about 20% more tracks
- CS jets look somewhat similar
- the efficiency increases similarly for HP and all tracks
wf150 0_tk_eff_pt
- fake rates are up for inclusive tracks (some cleanup is likely needed in a followup), while for HP tracks the fake rate is only about 10-20% (fractionally) higher
wf150 0_tk_fr_pt
- the duplicate rate is up significantly (is merging good enough or needs some reconfiguration eventually?)
wf150 0_tk_dup_pt

I'm ready to signoff pending the resolution of the hiRegitMuMixedTripletStepSeedLayersA configuration

File: RecoHI/HiTracking/python/hiMixedTripletStep_cff.py (hunk 20 above)
Comment (abaty) [reply RecoHI/HiTracking/python/hiMixedTripletStep_cff.py]: @slava77 I think having the extra layers in the list for the Regit step is fine; these steps tend to run quite quickly and produce a fairly small number of tracks.  It is probably best for consistency to keep these two iterations using the same layer list.

**Comment by fabiocos** [comment]
+1
